source/2_platform/rex_opengl/src/resources/buffer_layout.cpp

```diff
@@ -4,170 +4,132 @@
-        if (*this == other)
-        {
-            return *this;
-        }
-
-        m_elements = other.m_elements;
-
-        return *this;
-    }
-    //-----------------------------------------------------------------------
-    BufferLayout& BufferLayout::operator=(BufferLayout&& other) noexcept
-    {
-        if (*this == other)
-        {
-            return *this;
-        }
-
-        m_elements = rsl::exchange(other.m_elements, {});
-
-        return *this;
-    }
-
-    //-----------------------------------------------------------------------
-    BufferLayout::BufferLayout(const BufferElements& elements)
-        :m_elements(elements)
-    {
-
-    }
-
-    //-----------------------------------------------------------------------
-    bool BufferLayout::empty() const
-    {
-        return m_elements.empty();
-    }
-    //-----------------------------------------------------------------------
-    s32 BufferLayout::size() const
-    {
-        return m_elements.size();
-    }
-
-    //-----------------------------------------------------------------------
-    const BufferElements& BufferLayout::get_buffer_elements() const
-    {
-        return m_elements;
-    }
-
-    //-----------------------------------------------------------------------
-    bool operator==(const BufferLayout& layout1, const BufferLayout& layout2)
-    {
-        if (layout1.get_buffer_elements().size() != layout2.get_buffer_elements().size())
-        {
-            return false;
-        }
-
-        return layout1 == layout2;
-    }
-    //-----------------------------------------------------------------------
-    bool operator!=(const BufferLayout& layout1, const BufferLayout& layout2)
+  }
+
+  //-------------------------------------------------------------------------
+  bool operator==(const BufferElement& element1, const BufferElement& element2)
+  {
+    return element1.type == element2.type && element1.component_count == element2.component_count && element1.normalized == element2.normalized;
+  }
+
+  //-------------------------------------------------------------------------
+  bool operator!=(const BufferElement& element1, const BufferElement& element2)
+  {
+    return !(element1 == element2);
+  }
+
+  //-----------------------------------------------------------------------
+  BufferLayoutBuilder::BufferLayoutBuilder()
+    :m_layout({})
+  {
+
+  }
+  //-----------------------------------------------------------------------
+  BufferLayout& BufferLayoutBuilder::add_buffer_element(DataType inType, ShouldNormalize inShouldNormalize /*= ShouldNormalize::no*/)
+  {
+    BufferElements new_buffer_elements;
+
+    new_buffer_elements.reserve(m_layout.size() + 1);
+    new_buffer_elements.insert(rsl::end(new_buffer_elements), rsl::begin(m_layout.get_buffer_elements()), rsl::end(m_layout.get_buffer_elements()));
+    new_buffer_elements.emplace_back(BufferElement(inType, buffer_layout::get_data_type_component_count(inType), inShouldNormalize));
```

Comment:
with emplace_back, you can drop the BufferElement constructor call, just pass in the arguments you'd pass to the constructor